kong/globalpatches.lua

```diff
@@ -589,6 +591,31 @@ return function(options)
       return sock
     end
 
+    -- OTel-formatted logs feature
+    local dynamic_hook = require "kong.dynamic_hook"
+    local hook_called = false
+    _G.ngx.log = function(...)
+      if hook_called then
+        -- avoid recursive loops: relies on the hook handler to NOT yield
+        old_ngx_log(ngx.ERR, debug.traceback("concurrent execution detected for: ngx.log", 2))
+        return old_ngx_log(...)
+      end
+
+      -- stack level = 5:
+      -- 1: maybe_push
+      -- 2: dynamic_hook.pcall
+      -- 3: dynamic_hook.run_hook
+      -- 4: patched function
+      -- 5: caller
+      hook_called = true
+      dynamic_hook.run_hook("observability_logs", "push", 5, ...)
+      hook_called = false
```

Comment:
thanks for looking into this extensively @ADD-SP, I had a chat about this with @bungle the other day, who pointed out that if ever `ngx.log` started yielding we would experience issues even outside of this particular feature (ngx.log by definition can run almost anywhere, but yielding is not always possible, etc). For this reason I think we should be careful with trying to cover up the yielding behavior and still "make things work".

We could go as far as considering to just error out loudly if we detect this (i.e. when `hook_called` is true and `ngx.log` is called again). That said, we now have [a test](https://github.com/Kong/kong/blob/ad0b93cd99f5d3badd6123342979051f90c998da/spec/02-integration/14-observability/05-logs_spec.lua) that should prevent us introducing a yielding behavior in `ngx.log` by any means and I think we should just work on ensuring that test is correct while keeping the logic here as simple as possible.

In the edge case scenario where some bug in the hook itself throws an error outside of the protected pcall execution, I think it's fine that we end up in the branch that only executes the old `ngx.log` function without OTel logs (current implementation). If that ever happens it means we'd have some bug to fix in the hooks, that's also something we don't want to cover up.

Let me know what you think (@bungle your opinion would also be appreciated :) ).

Reply:
> For this reason I think we should be careful with trying to cover up the yielding behavior and still "make things work".

@samugi Agreed, also, I don't like to cover up such errors, so I think we could loudly log errors and then "make things work".

Making things work after logging errors loudly could give a time window to support people, if customers consider failure log shipping to be a critical bug, auto-recovery is more friendly than restarting manually for customers, the more friendly mitigation, the less possibility for an escalation.

An escalation always interrupts something and takes the bandwidth of other tasks, so I just want to reduce the possibility of it.

Reply:
In the fault tree analysis, we cover all possible faults, all fault branches can recover by themself, no need to restart the worker process, and there is only one fault branch we cannot log errors loudly, if we can detect this case and log something, we will never cover up anything.

Reply:
> That said, we now have [a test](https://github.com/Kong/kong/blob/ad0b93cd99f5d3badd6123342979051f90c998da/spec/02-integration/14-observability/05-logs_spec.lua) that should prevent us introducing a yielding behavior in ngx.log by any means and I think we should just work on ensuring that test is correct while keeping the logic here as simple as possible.

Maybe I'm worrying too much, but introducing yielding accidentally for some edge cases is possible in the future, and it is hard to cover all branches of [`maybe_push`](https://github.com/Kong/kong/blob/ad0b93cd99f5d3badd6123342979051f90c998da/kong/observability/logs.lua#L103) using test cases, so introducing timeout mechanism might be a safeguard for future, and error logs make sure we will not cover-up such errors.

Just like I said above, giving our support people a comfortable time window to fix that, escalation is always hard.

Reply:
I agree with @samugi that additional defenses are not needed.

Reply:
@samugi @hanshuebner, let's put the discussion of additional defenses later.

Could we log errors loudly if the `hook_called` is keeping `true`, just to make sure we are not failing silently? Customers might not notice this failure quickly, and open a support case after a long time, this makes troubleshooting more difficult since the time windows might be so long.

Reply:
we are logging with `error` level here: https://github.com/Kong/kong/blob/be23116c59abfee82d284024a9e5f0424a8d74ac/kong/globalpatches.lua#L599-L600
Is this not enough?

Reply:
> we are logging with `error` level here:
> 
> https://github.com/Kong/kong/blob/be23116c59abfee82d284024a9e5f0424a8d74ac/kong/globalpatches.lua#L599-L600
> 
> Is this not enough?

@samugi Ah, sorry, I missed this line, I think this error should loudly enough.

Reply:
I added a (currently redundant) pcall to be absolutely sure the hook call cannot error, hopefully that addresses the remaining concerns

Reply:
Let's unblock this PR, it is not worth blocking this PR for too long just because of the additional defenses.

Reply:
We can discuss it after merging.